app/src/org/commcare/tasks/ProcessAndSendTask.java

```diff
@@ -174,60 +177,65 @@ private boolean checkFormRecordStatus(FormRecord[] records)
 
             //If the form is complete, but unprocessed, process it.
             if (FormRecord.STATUS_COMPLETE.equals(record.getStatus())) {
+                SQLiteDatabase db =
+                        CommCareApplication.instance().getUserDbHandle();
+                db.beginTransaction();
                 try {
                     records[i] = processor.process(record);
+                    db.setTransactionSuccessful();
                 } catch (InvalidStructureException | XmlPullParserException |
                         UnfullfilledRequirementsException e) {
-                    handleExceptionFromFormProcessing(record, e);
+                    db.endTransaction();
+                    records[i] = handleExceptionFromFormProcessing(record, e);
                     needToSendLogs = true;
                 } catch (FileNotFoundException e) {
+                    db.endTransaction();
                     if (CommCareApplication.instance().isStorageAvailable()) {
                         //If storage is available generally, this is a bug in the app design
                         Logger.log(LogTypes.TYPE_ERROR_DESIGN,
                                 "Removing form record because file was missing|" + getExceptionText(e));
                         record.logPendingDeletion(TAG,
                                 "the xml submission file associated with the record could not be found");
                         FormRecordCleanupTask.wipeRecord(c, record);
+                        records[i] = FormRecord.StandInForDeletedRecord();
                     } else {
                         CommCareApplication.notificationManager().reportNotificationMessage(
                                 NotificationMessageFactory.message(ProcessIssues.StorageRemoved), true);
                         //Otherwise, the SD card just got removed, and we need to bail anyway.
                         throw e;
                     }
                 } catch (IOException e) {
+                    db.endTransaction();
                     Logger.log(LogTypes.TYPE_ERROR_WORKFLOW, "IO Issues processing a form. " +
                             "Tentatively not removing in case they are resolvable|" + getExceptionText(e));
+                } catch (Exception e) {
+                    db.endTransaction();

```

Comment:
Why are we adding this? To ensure that we close the database handle regardless? If so seems like we should log this rather than fail silently on any non-caught exception

Reply:
Ah wow, good catch. I think what I should probably do is just preserve the old behavior precisely, by re-throwing the exception after ending the transaction. Does that sound right to you?

Reply:
:+1: